### ISSUE TITLE
Do not delete Qt buildtrees and packages folders! Qt no longer works.

### ISSUE THREAD

**Issue by claudiofantacci** [Microsoft/vcpkg#3248]
Qt dlls have hardcoed build tree paths in the dll (you can easily check this with an HEX editor).

Do not delete any Qt buildtrees folder, as suggested in the [vcpkg FAQ](https://vcpkg.readthedocs.io/en/latest/about/faq/) to free up some space, or you will break your setup.

If you delete such folders (I did it to free up to 100 GB of space), dlls will no longer be able to find/open/use plugins.

**Comment by claudiofantacci** [comment]
> qt apps have baked-in-at-compile-time paths that they default back on to when no qt.conf is provided. That's what you see in a hex editor.

This implies that no `qt.conf ` is present in the vcpkg installation of qt5. Am I right?
I saw #2908 where you posted about qt.conf files and I tired to ad a conf manually without success. However it seems to me that #2908 is related to tools, does this also mean that several conf file are missing from for the vcpkg installation?

**Comment by glachancecmaisonneuve** [comment]
> This implies that no qt.conf is present in the vcpkg installation of qt5. Am I right?

yes. that is correct.  there is a qt_debug.conf and a qt_release.conf that give passable values.  One or the other (depending on your build) should be renamed to qt.conf and PREFIX replaced with ../..

just to be clear, this wont fix the cmake macros.  it's going to allow qdesigner, qmake and other tools to find their libraries correctly

**Comment by glachancecmaisonneuve** [comment]
I was not sure what you meant by 

> dlls have hardcoed build tree paths in the dll 

I figured you meant that some tools were trying to link to dlls in buildtrees.

**Comment by claudiofantacci** [comment]
Let me reply in order 😃 

> yes. that is correct. there is a qt_debug.conf and a qt_release.conf that give passable values. One or the other (depending on your build) should be renamed to qt.conf and PREFIX replaced with ../..

I see. However vcpkg installs both debug and release: both files are copied, without being renamed and PREFIX filled, in the `tools` folder. What should we do then? Sorry to ask again: is there a conf file for each DLL? Could you please provide some details (or reference) about this?

> I was not sure what you meant by
> 
> > dlls have hardcoed build tree paths in the dll
> 
> I figured you meant that some tools were trying to link to dlls in buildtrees.

Long story short: I deleted the `buildtrees` folder and some qt DLL where complaining they could not find a source file (yes, source) within the `buildtrees` folder. I opened the DLL with an HEX editor and found out that those path where hardcoded within the DLL.
Based on what you said, _without_ conf files, the DLLs fallback to some hardcoded path, otherwise they shall work properly. As a consequence, this implies that the vcpkq installation of qt5 is missing an important step of _installing and configuring_ such configuration files.

**Comment by glachancecmaisonneuve** [comment]
The part I don't understand:

> DLL where complaining

I know i'm probably being very dense but DLLs don't do anything.  They are loaded and executed within an executable.  I'm really not trying to be pedantic I'm dense.

If the exe inherits from QApplication (like qmake and qdesigner do), then a qt.conf file, installed in the same folder the exe is in, will inform the QApplication as to where to find what it needs to run.  The way that works is that all QApplications have the same initialization routine that creates an instance of QLibrary.  QLibrary, in turn, looks if there is a qt.conf to read in the apps folder, and then magic happens.

**Comment by claudiofantacci** [comment]
@glachancecmaisonneuve you're right, sorry for insisting on the DLL part and now I get your point.

The executable I was running was in another installation folder and at runtime was complaining about some missing file needed by a Qt DLL.  
The application I'm talking about is not developed by me, but by some other people of my lab. I'll point them to here so that we can understand this point and properly combat the problem.

Thanks!

**Comment by drdanz** [comment]
Using the qt.conf file works, but unfortunately, this also means that any project should install a qt.conf file in the binary directory, that is not needed for any other configuration:
* Official binaries installed from Qt installers for windows and for all the other platforms do not need a qt.conf file
* Binaries installed by other operating systems (linux, macos) do not need a qt.conf file

vcpkg is the only "platform" that requires a qt.conf file, and this means that every developer will have to add a qt.conf file in their packages, just to support vcpkg. This is never going to happen there will always be packages that don't work with vcpkg because they don't have the qt.conf file

My suggestion is to patch the installed files with the actual installation path using the qt installer framework patch operation as suggested in the second answer here: https://stackoverflow.com/questions/4699311/how-to-install-qt-on-windows-after-building

**Comment by glachancecmaisonneuve** [comment]
> My suggestion is to patch the installed files with the actual installation path using the qt installer framework patch operation.

This is already done, through qtdeploy.ps1.

Rereading myself I can see that I am implying that this qt.conf file needs to be installed in any app built with vcpkg.  This is wrong.  It is only true for apps installed **in** the vcpkg tree, like qmake and qdesigner.

Once you are done building an app, you should run qtdeploy.ps1 that will copy all the necessary dlls to the apps folder.  qtdeploy.ps1 is a replacement for winqtdeploy.

**Comment by glachancecmaisonneuve** [comment]
@claudiofantacci I realize that I may have pointed you in the wrong direction.  

If you have an app that need to be independent of it's build system, something you intend redistributing for example, then you should try the qtdeploy.ps1 script.  It should detect and copy all the dependencies your app needs in the app's folder, and your app will be entirely independent of it's build system.  qtdeploy.ps1 is an internal script developed by ras that replaces the usual deploy system, called winqtdeploy.

**Comment by claudiofantacci** [comment]
Thanks @glachancecmaisonneuve for your comments and for pointing me to `qtdeploy.ps1`.
I still think that there is something going wrong in the vcpkg build of Qt5.

Long story short: I compile a library and some applications that uses Qt5 using the vcpkg cmake toolchain. When Qt5 was a single portfile (version 5.8.6) I had no problem running such applications from either their bin build or installation directory, without running any script or using any configuration file. Updating to Qt 5.9.2 the Qt-based application where not able to run anymore. From that point we arrive here and the only way we find out to run our applications again is to put a `qt.conf` file in out bin directory.
By my understanding, in this settings, I don't have to run `qtdeploy.ps1` (and it has never be the case). Am I missing something here? Probably updating the port version something wrong happened?

**Comment by glachancecmaisonneuve** [comment]
(Incoming wild guess).  Have you tried adding vcpkg/installed/<triplet>/bin to your path?  

If you are just building against qt (not building something that needs integration like a qdesigner plugin), and it's not time for deploying yet, adding vcpkg/<triplet>/bin to your path it's often all you need.  

Sidenote, if you were to use qt's official distribution for building, the path is set by the qtenv.bat file that qt developers know to call before building.  The difference is that since the official qt tools put everything in the same folder (the tools and the dlls and all in the same location) while vcpkg separates the tools from the dlls.

**Comment by claudiofantacci** [comment]
> (Incoming wild guess)

They are super welcome, thanks!

> Have you tried adding vcpkg/installed//bin to your path?
> 
> If you are just building against qt (not building something that needs integration like a qdesigner plugin), and it's not time for deploying yet, adding vcpkg//bin to your path it's often all you need.

I have the `vcpkg/bin` into my PATH.  
My guess is that something changed from port version 5.8 (which was single package, but it hardly make any difference) to 5.9. In particular, I think that the fallback paths to the buildtrees folder has been either changed (unlikely), deleted (even more unlikely) or not properly set (which may be the case, I could not find any path opening the DLLs with an HEX editor). I can only get my Qt-based application running with a manual `qt.conf` file.

As suggested by @drdanz, we should consider patching the installed files with the actual installation path using the Qt installer framework patch. This should be done because vcpkg installs (hence moves) Qt from its build folder and the absolute paths within the DLL are bound to the buildtrees directory that may be deleted. This also imply, that something changed/broke from port version 5.8 to 5.9.

**Comment by claudiofantacci** [comment]
A quick update on this issue.

I modified the `qt5-base` portfile in my [forked vcpkg repository](https://github.com/claudiofantacci/vcpkg/commit/3b7f1bd6fc07c537ecbb2256f2df5358b396ad6f) in order to have the proper **hardcoded fullpath** in Qt5 DLLs.  
The change is under testing on our systems. As a result of this new portfile, developers are no longer required to provide a `qt.conf` file, as it should be for compiled libraries.

There is still a majour issue incurring with `Qt5` and `vcpkg`: should you export a package that uses `Qt5` and anyone install it on a local machine, the `Qt5` DLLs shall be patched to account for the new installation folder (_this is exactly what the `Qt5` installare does for pre-compiled binaries_).
Either `vcpkg` or `Qt5` should find a way to combat this issue.

**Comment by claudiofantacci** [comment]
> I wonder how this need to patch the Qt5 DLLs interacts with the vcpkg export

This patch is needed because there are wrong absolute paths in `Qt5Core.dll`. In particular, once `qt5-base` module is compiled it is then installed in the `packages` folder and the DLL will have that _absolute path_ hardcoded. vcpkg will then copy the DLLs in the `installed` folder and here things break because during copy the absolute paths are not patched.

You can see this whole procedure as if your cross-compiling `Qt5` having the same host and the target PC. The moment you move the compiled DLLs from the host to the target PC you have to patch the hardcoded paths in the DLLs.

If you want to see this, just open `Qt5Core.dll` with an HEX editor and look for `qt_prfxpath` (or simply the prefix path of vcpkg).

> in particular when producing `ifw` installers

I honestly was unaware of this feature that is actually the best we can have.
If `QtIFW` works well, it should be able to patch the DLLs in the same way the official Qt installer does.  
The patch I propose changes only a path which can be found in the DLL (as explained before), thus everything that comes after, including any `export` command should just work the same.

**Comment by podsvirov** [comment]
In general, we need to fix the library files after the installation.
I have an assumption that in Qt there is a utility that solves this problem, but I do not know which one. Undoubtedly, the solution is present in QtSDK. If I have time for this, I'll try to study their code base and suggest a solution.

**Comment by claudiofantacci** [comment]
@podsvirov, thanks for reaching out!
AFAIK/U there is [windeployqt](https://doc.qt.io/qt-5/windows-deployment.html) that should be a product of the compilation of `Qt`. However, I'm not able to find it.  
Also, if you use the `QtIFW` you should already have this "feature" built-in the installer.

**Comment by claudiofantacci** [comment]
> @claudiofantacci, QtIFW itself is an abstract tool. In general, it has nothing to do with installing Qt.

I'm not an expert about this, nor I have previous experience about this in particular.
I was expecting that [QtIFW](https://doc.qt.io/qtinstallerframework/ifw-overview.html) was providing a tool to fix the hardcoded paths. At the same time I was quite confused by the presence of [windeployqt](https://doc.qt.io/qt-5/windows-deployment.html).
At this point I'm guessing that I misunderstood the scope of former tool and maybe the appropriate one for us is the latter.

Let me know what you think about it so that maybe I can help you out 😄

**Comment by claudiofantacci** [comment]
I would just like to ping on this issue that also has PR #3280 associated with it.

It would be nice to understand whether the maintainer are aware of this issue with `Qt` and `vcpkg` or not. To me, this is quite an important matter, expecially if you want to distribute the binary of an application.

**Comment by claudiofantacci** [comment]
I'd like to reopen this issue after some of my colleagues and I are facing a problem with `Qt5`.
Basically we compile some GUI applications linking against `Qt5` compiled with `vcpkg` and our products either do not run or shows up with no icons. The reasons behind the failers are detailed in https://github.com/robotology/robotology-superbuild/issues/134#issuecomment-440063486 by @traversaro. This issue was present in the past and I fixed in with #3280, that was partially reverted in Microsoft/vcpkg@473d63c.

I try to explain where I think the problem is now. Please bear with me as I'm not a `Qt` expert 😄 

If you open the `Qt5Core.dll` with an hex-editor you will get something like the following list somewhere in the file:
```
qt_prfxpath=C:/Users/cfantacci/vcpkg/installed/x64-windows

...

C:/Users/cfantacci/vcpkg/packages/qt5-base_x64-windows/share/qt5/doc
include
lib
C:/Users/cfantacci/vcpkg/packages/qt5-base_x64-windows/share/qt5/bin
bin
C:/Users/cfantacci/vcpkg/packages/qt5-base_x64-windows/plugins
C:/Users/cfantacci/vcpkg/packages/qt5-base_x64-windows/share/qt5/imports
C:/Users/cfantacci/vcpkg/packages/qt5-base_x64-windows/qml
C:/Users/cfantacci/vcpkg/packages/qt5-base_x64-windows/share/qt5
C:/Users/cfantacci/vcpkg/packages/qt5-base_x64-windows/share/qt5
C:/Users/cfantacci/vcpkg/packages/qt5-base_x64-windows/share/qt5/translations
C:/Users/cfantacci/vcpkg/packages/qt5-base_x64-windows/share/qt5/examples
tests 
```
Now, by my understanding, every paths that is used by `Qt` start with `qt_prfxpath` and is expanded with the other entries of the list. Not sure if the expantion happens only to relative paths, but I would say so.
If I look at the directory `C:\Users\cfantacci\vcpkg\packages\qt5-base_x64-windows` I have
```
Mode                LastWriteTime         Length Name
----                -------------         ------ ----
d-----        18-Dec-18     12:49                bin
d-----        18-Dec-18     12:49                debug
d-----        18-Dec-18     12:33                include
d-----        18-Dec-18     12:50                lib
d-----        18-Dec-18     12:50                plugins
d-----        18-Dec-18     12:50                share
d-----        18-Dec-18     12:49                tools
-a----        18-Dec-18     12:50             46 BUILD_INFO
-a----        18-Dec-18     12:50            331 CONTROL
```
for the directory `C:\Users\cfantacci\vcpkg\packages\qt5-base_x64-windows\share`
```
Mode                LastWriteTime         Length Name
----                -------------         ------ ----
d-----        18-Dec-18     12:49                cmake
d-----        18-Dec-18     12:49                qt5
d-----        18-Dec-18     12:50                qt5-base
```
for the directory `C:\Users\cfantacci\vcpkg\packages\qt5-base_x64-windows\share\qt5`
```
Mode                LastWriteTime         Length Name
----                -------------         ------ ----
d-----        18-Dec-18     12:32                debug
d-----        18-Dec-18     12:49                doc
d-----        18-Dec-18     12:49                mkspecs
```
and finally for the directory `C:\Users\cfantacci\vcpkg\packages\qt5-base_x64-windows\share\qt5-base`
```
Mode                LastWriteTime         Length Name
----                -------------         ------ ----
-a----        13-Sep-18      6:25           8173 copyright
```
Looking back at our list, **the following paths are wrong**:
```
C:/Users/cfantacci/vcpkg/packages/qt5-base_x64-windows/share/qt5/imports
C:/Users/cfantacci/vcpkg/packages/qt5-base_x64-windows/qml
C:/Users/cfantacci/vcpkg/packages/qt5-base_x64-windows/share/qt5/translations
C:/Users/cfantacci/vcpkg/packages/qt5-base_x64-windows/share/qt5/examples
```
This paths can be overridden by the following `qt.conf`
```ini
[Paths]
Prefix = C:/Users/cfantacci/vcpkg/installed/x64-windows
Documentation = share/qt5/doc
Headers = include
Libraries = lib
Binaries = tools/qt5
LibraryExecutables = tools/qt5
Plugins = plugins
Qml2Imports = qml
Data = share/qt5
ArchData = share/qt5
HostData = share/qt5
HostBinaries = tools/qt5
```
to the correct folders. In particular, I guess the `qml` folder is the one that is causing our issues.
I'm not 100% sure that this is the real cause of the failure, but for sure there are some mistakes that are worth taking into account.

I think that the error may lie in the following reverted paths: https://github.com/Microsoft/vcpkg/commit/473d63c#diff-ae2ab5cd22202159887c7d80ea952a4e. It may be worth trying to revert the reverted commit and see how it goes 😄 What do you think about it?
